### PR TITLE
Refactor google map views - 1/x

### PR DESCRIPTION
**Pull request by jankrzempek**
https://trello.com/c/gUta3Mcr/706-🛠-refactor-split-google-map-class-to-smaller-one-based-on-the-use-case

### CODE REVIEW

File: AirCasting/Map/GoogleMap/GoogleMapButler.swift [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    
+    var cameraUpdate: GMSCameraUpdate {
+        let initialBounds = GMSCoordinateBounds()
+        guard liveModeOn else {
+            let pathPointsBoundingBox = pathPoints.reduce(initialBounds) { bounds, point in
+                bounds.includingCoordinate(point.location)
+            }
+            return GMSCameraUpdate.fit(pathPointsBoundingBox, withPadding: 1.0)
+        }
+        let pathPointsBoundingBox = initialBounds.includingCoordinate(pathPoints.last!.location)
```
Comment (anna1901) [inline AirCasting/Map/GoogleMap/GoogleMapButler.swift:47]: You should make sure that pathPoints isn't empty

File: AirCasting/Map/GoogleMap/GoogleMapView.swift [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+            mapView.moveCamera(butler.cameraUpdate)
+            if mapView.camera.zoom > 16 {  mapView.animate(toZoom: 16) }
+            // Centers the user onto last recorded position
+        }
+        
+        return mapView
+    }
+    
+    func updateUIView(_ uiView: GMSMapView, context: Context) {
+        context.coordinator.butler = butler
```
Comment (anna1901) [inline AirCasting/Map/GoogleMap/GoogleMapView.swift:76]: Why do we need to assign butler on updateUIView?

File: AirCasting/Map/GoogleMap/GoogleMapView.swift [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+        
+        if butler.notesNumber != context.coordinator.mapNotesCounter {
+            butler.placeNotes(uiView, context: context)
+            uiView.isMyLocationEnabled = false
```
Comment (anna1901) [inline AirCasting/Map/GoogleMap/GoogleMapView.swift:80]: Why is this set to false?

File: AirCasting/Map/GoogleMap/GoogleMapView.swift [hunk 4]
```diff
@@ -0,0 +1,184 @@
+//  Created by Lunar on 26/01/2021.
+//
+import UIKit
+import SwiftUI
+import GoogleMaps
+import Resolver
+
+struct GoogleMapView: UIViewRepresentable {
```
Comment (anna1901) [inline AirCasting/Map/GoogleMap/GoogleMapView.swift:8]: Is this map only used to display session location and path? Because if yes, then maybe we could rename it to indicate that? Because I see that it is tightly associated with notes and heatmap logic, so it's not really a generic GoogleMap

**Review by anna1901 — APPROVED** [review]
Some maps don't have the right style in the darkmode (on placepicker screen when creating mobile session and on session card), but I'm giving approve anyway because other than that everything is working well! Congrats! I know this required a lot of work. 🚀

File: AirCasting/Map/GoogleMap/GoogleMapView.swift [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+    @ObservedObject private var butler: GoogleMapButler
+    
+    @Binding private var noteMarketTapped: Bool
```
[the remark this answers is not in the document]
Comment (jankrzempek) [reply AirCasting/Map/GoogleMap/GoogleMapView.swift:11]: 😅 aj, thanks!

File: AirCasting/Map/GoogleMap/GoogleMapView.swift (hunk 3 above)
Comment (jankrzempek) [reply AirCasting/Map/GoogleMap/GoogleMapView.swift:80]: I think it should not be here at all, some leftover - thx!

File: AirCasting/Map/GoogleMap/GoogleMapView.swift (hunk 2 above)
Comment (jankrzempek) [reply AirCasting/Map/GoogleMap/GoogleMapView.swift:76]: To keep it up to date and be sure that data transfers on every map update - this is why. I will try to explain it further in the upcoming days as I will gain more knowledge which I would like to get first.

File: AirCasting/Map/GoogleMap/GoogleMapView.swift (hunk 4 above)
Comment (jankrzempek) [reply AirCasting/Map/GoogleMap/GoogleMapView.swift:8]: Yeah, you are right. The generic approach here is used (somehow) in GoogleMapCore which I think can be the base for every other map we will be using/creating (as this one). This map as you said is used for many things - saying that I don't have idead for a better name to be honest. Would like like to still rename it? If so, do you have maybe some ideas?

File: AirCasting/Map/GoogleMap/GoogleMapButler.swift [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+    func initialMakeUIVIew(context: UIViewRepresentableContext<GoogleMapView>) -> GMSMapView {
+        let mapView = coreMap.initialCommonMakeUIView()
+        mapView.settings.myLocationButton = showMyLocationButton
+        placeNotes(mapView, context: context)
+        drawPolyline(mapView, context: context)
+        mapView.isMyLocationEnabled = isMyLocationEnabled
+        return mapView
+    }
+    
+    func applyStylying(to mapView: GMSMapView) { coreMap.apllyStylying(to: mapView) }
```
Comment (pawelgil) [inline AirCasting/Map/GoogleMap/GoogleMapButler.swift:65]: this is a second `applyStyling` I see, but this one is not related to the `CoreGoogleMap` protocol. If's kind of confusing what does what

File: AirCasting/Map/GoogleMap/GoogleMapButler.swift [hunk 7]
```diff
@@ -0,0 +1,154 @@
+// Created by Lunar on 27/07/2022.
+//
+import Foundation
+import SwiftUI
+import GoogleMaps
+import Resolver
+
+class GoogleMapButler: ObservableObject {
```
Comment (pawelgil) [inline AirCasting/Map/GoogleMap/GoogleMapButler.swift:8]: you'll need to explain to me what is the idea behind those butler objects

**Review by pawelgil — CHANGES_REQUESTED** [review]
It looks quite unstructured to me for a refactoring. Can you please provide some simple overview of how it's going to look like in the end?

**Comment by jankrzempek** [comment]
> Please for such changes provide an extensive explanation in the PRs description (what's the refactoring plan? How is it achieved, what still needs to be done?)

_Scheme (I am not sure if it is right, but I am going to use the below explain to make it clear):_
Zrzut ekranu 2022-08-9 o 15 56 56

**Plan:**
- Restructure/refactor GoogleMapView - every map which refers to `GoogleMaps`

**Plan for this PR**

- Refactor `GoogleMapView` struct

**Refactoring Plan**
- Delete all the logic which was obsolete/not needed
- Think about splitting this struct into smaller/easier-to-maintain objects

**Refactoring Plan**
- Go through every map that we have and decide what part can be classified as a common one (those which can be, move then to `CoreMap`)
- Create GoogleMap for choosing a custom location
- Having the above ready, move to clear/remove unused code from the GoogleMapView struct
- Both "new" structs should be equipped with their own Butlers (Naming comes from the Twitter conversation where people decided it can be something similar to controller - it should handle some part of view logic behind the scenes. I really liked the idea, and for me, it sounded appropriate - that was the reason why I used it) 
- Every Butler should be able to use `GoogleMapCore`

**What still needs to be done**
- After applying all the comments and deciding that this PR is merge-ready, we should cover every other GoogleMap that we have inside our app
- Applying this refactor concept, we should equip every map with its `Butler` and think about its functionality (maybe splitting them can be a good idea?)

File: AirCasting/Map/GoogleMap/GoogleMapView.swift [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+    private(set) var threshold: SensorThreshold?
+    private var onPositionChange: (([PathPoint]) -> ())? = nil
+    
+    init(pathPoints: [PathPoint],
+         threshold: SensorThreshold? = nil,
+         isMyLocationEnabled: Bool = false,
+         isSessionActive: Bool = false,
+         isSessionFixed: Bool = false,
+         noteMarketTapped: Binding<Bool> = .constant(false),
+         noteNumber: Binding<Int> = .constant(0),
```
[the remark this answers is not in the document]
Comment (jankrzempek) [reply AirCasting/Map/GoogleMap/GoogleMapView.swift:27]: There are no notes in `CoreMap`, only in this view and its butler which supports it. Did you mean not to use notes at all?

File: AirCasting/Map/GoogleMapCore.swift [hunk 9]
```diff
@@ -0,0 +1,70 @@
+// Created by Lunar on 27/07/2022.
+//
+import SwiftUI
+import GoogleMaps
+import Resolver
+
+protocol GoogleMapCore {
```
[the remark this answers is not in the document]
Comment (jankrzempek) [reply AirCasting/Map/GoogleMapCore.swift:7]: In the beginning, I was thinking about making it easier to test - that was the main idea. Looking at it now, I am not sure if it will help.

File: AirCasting/Map/GoogleMapCore.swift [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    func initialCommonMakeUIView() -> GMSMapView 
+    func apllyStylying(to mapView: GMSMapView)
+    func optionalTrackerLocation() -> (Double, Double)
```
[the remark this answers is not in the document]
Comment (jankrzempek) [reply AirCasting/Map/GoogleMapCore.swift:10]: To let know the other coder that `tracker.location.value?` can be nil

File: AirCasting/Map/GoogleMapCore.swift [hunk 11]
```diff
@@ -0,0 +1,70 @@
+// Created by Lunar on 27/07/2022.
+//
+import SwiftUI
+import GoogleMaps
+import Resolver
+
+protocol GoogleMapCore {
+    func initialCommonMakeUIView() -> GMSMapView 
+    func apllyStylying(to mapView: GMSMapView)
```
[the remark this answers is not in the document]
Comment (jankrzempek) [reply AirCasting/Map/GoogleMapCore.swift:9]: Name changes. It was all about using JSON styling the map

File: AirCasting/Map/GoogleMap/GoogleMapButler.swift (hunk 6 above)
Comment (jankrzempek) [reply AirCasting/Map/GoogleMap/GoogleMapButler.swift:65]: The idea was, that GoogleMapView can use only Butler methods so I was in need to repeat somehow this JSON styling from GoogleMapCore

**Comment by pawelgil** [comment]
Some artifacts from our refactoring kickoff meeting:
```Swift
        enum TrackingStyle {
            case none
            case user
            case latestPathPoint
            case wholePath
        }
        
        enum UserIndicatorStyle {
            enum DotStyle {
                case standard //blue, google standard
                case airCasting(color: Color)
            }

            case none
            case live(style: DotStyle)
        }
        
        Map(trackingStyle: .none, indicatorStyle: .live(.standard))
            .onAppear({ map.updateLocation(location) })
```

```
// 1. Session type (static) - IN
// 2. Notes (marker tapped, note number, map notes)
//   - Marker tapped (dynamic) - OUT
//   - Note number (dynamic) - OUT. When marker-tapped is triggered it contains the note number
//   - Map notes (dynamic) - IN
// 3. is-user-intaracting - To remove (it was protecting the CPU, but not really)
// 4. placePicker - used only in `ChooseCustomLocationView`. Something to do with halting the map refresh (?)
// 5. threshold
// 6. pathPoints
```

image

**Comment by pawelgil** [comment]
Notes from last call:
```
enum MapViewType {
    case normal
    case satelite
    case terrain
    case hybrid
}

enum TrackingStyle {
    case none
    case user
    case latestPathPoint
    case wholePath
}

enum UserIndicatorStyle {
    enum DotStyle {
        case standard //blue, google standard
        case airCasting(color: Color)
    }
    
    case none
    case live
    case followingPath(color: Color)
}

// Map starting camera point: users location or (if nil) Apple park
       
// isMyLocationEnabled in init with default to `true`


//Map(trackingStyle: .none, indicatorStyle: .live(.standard))
//   .onAppear({ map.updateLocation(location) })

// superview properies: thresholds & session measurements
// @State currentColor: Color

// Map(trackingStyle: .latestPathPoint,
//     indicatorStyle: .followingPath(currentColor))
//     mapviewType: .hybrid,
//    .onAppear({ map.updateLocation(location) })


//ZStack {
//    Map()
//        .onCameraTranslation { coordinates in
//            camera = coordinates
//        }
//
//    HeatmapOverlay(camera: camera)
//}
```
